onnxruntime/contrib_ops/cpu/reverse.cc

```diff
@@ -0,0 +1,300 @@
+  AllocatorPtr allocator;
+  ctx->GetTempSpaceAllocator(&allocator);
+  ORT_ENFORCE(allocator != nullptr, "Temporary memory allocator failed");
+
+  const int64_t len = shape.Size();
+  ORT_ENFORCE(len > 0, "Need atleast one float16 element to be processed in Reverse operator");
+
+  // allocate intermediate buffers to be used for processing
+  float* input_buffer = static_cast<float*>(allocator->AllocArray(sizeof(float), len));
+  float* output_buffer = static_cast<float*>(allocator->AllocArray(sizeof(float), len));
+  ORT_ENFORCE(input_buffer && output_buffer, "Could not allocate enough memory " 
+	                                         "to process float16 data in Reverse operator");
+
+  const auto& dims = shape.GetDims();
+  const auto& eigen_reverse_axes = generate_bool_reverse_axes<rank>(reverse_axes);
+
+  // fill the intermediate input buffer with the values in the input tensor
+  const auto& span = gsl::make_span(input_tensor->Data<BFloat16>(), len);
+  std::transform(span.cbegin(), span.cend(), input_buffer, [](const BFloat16& val) { return val.ToFloat(); });
+
+  // process the intermediate input buffer and use the intermediate output buffer to hold the output
+  EigenTensorMap<float, rank> eigen_tensor_input = buffer_as_eigen_tensor<float, rank>(input_buffer, dims);
+  EigenTensorMap<float, rank> eigen_tensor_output = buffer_as_eigen_tensor<float, rank>(output_buffer, dims);
+
+  eigen_tensor_output = eigen_tensor_input.reverse(eigen_reverse_axes);
+
+  // fill output tensor's values with results in intermediate output buffer
+  auto* output_data = output_tensor->template MutableData<BFloat16>();
+  std::transform(output_buffer, output_buffer + len, output_data, [](const float& val) { return BFloat16(val); });
+
+  // free the intermediate buffers
+  allocator->Free(input_buffer);
+  allocator->Free(output_buffer);
+}
+
+#define ProcessType(type, ml_data_type, rank, dims)                                                              \
+  const ConstEigenTensorMap<type, rank> input =                                                                  \
+  buffer_as_const_eigen_tensor<type, rank>(static_cast<const type*>(input_tensor->DataRaw(ml_data_type)), dims); \
+  EigenTensorMap<type, rank> output =                                                                            \
+  buffer_as_eigen_tensor<type, rank>(static_cast<type*>(output_tensor->MutableDataRaw(ml_data_type)), dims);     \
+  output = input.reverse(eigen_reverse_axes);
+
+template <int rank>
+void ReverseImpl(const OpKernelContext* ctx, const Tensor* input_tensor, Tensor* output_tensor, 
+const TensorShape& shape, const std::vector<int64_t>& reverse_axes) {
+  const auto& dtype = input_tensor->DataType();
+  const auto& dims = shape.GetDims();
+  const auto& eigen_reverse_axes = generate_bool_reverse_axes<rank>(reverse_axes);
+
+  if (dtype == DataTypeImpl::GetType<float>()) {
```

Comment:
>DataTypeImpl::GetType<float>() [](start = 15, length = 30)

Expensive query

Reply:
You mean checking the type is an expensive query ? We need to know the type at that point to proceed, don't we ? What's the alternate that you have in mind to cross this hurdle ?